src/engine/TransitivePathBase.cpp

```diff
@@ -36,29 +37,54 @@ TransitivePathBase::TransitivePathBase(
   AD_CORRECTNESS_CHECK(qec != nullptr);
   AD_CORRECTNESS_CHECK(subtree_);
   if (lhs_.isVariable()) {
-    variableColumns_[std::get<Variable>(lhs_.value_)] =
-        makeAlwaysDefinedColumn(0);
+    variableColumns_[lhs_.value_.getVariable()] = makeAlwaysDefinedColumn(0);
   }
   if (rhs_.isVariable()) {
-    variableColumns_[std::get<Variable>(rhs_.value_)] =
-        makeAlwaysDefinedColumn(1);
+    variableColumns_[rhs_.value_.getVariable()] = makeAlwaysDefinedColumn(1);
   }
-  if (minDist_ == 0 && lhs_.isUnboundVariable() && rhs_.isUnboundVariable()) {
-    emptyPathBound_ = true;
-    lhs_.treeAndCol_.emplace(makeEmptyPathSide(qec, std::move(activeGraphs)),
-                             0);
+  if (minDist_ == 0) {
+    auto& startingSide = decideDirection().first;
+    // If we have hardcoded differing values left and right, we can increase the
+    // minimum distance to 1
+    if (!lhs_.isVariable() && !rhs_.isVariable() &&
+        lhs_.value_ != rhs_.value_) {
+      minDist_ = 1;
+    } else if (lhs_.isUnboundVariable() && rhs_.isUnboundVariable()) {
+      boundVariableIsForEmptyPath_ = true;
+      lhs_.treeAndCol_.emplace(makeEmptyPathSide(qec, std::move(activeGraphs)),
+                               0);
+    } else if (!startingSide.isVariable()) {
+      startingSide.treeAndCol_.emplace(
+          joinWithIndexScan(qec, std::move(activeGraphs), startingSide.value_),
+          0);
+    }
   }
 
   lhs_.outputCol_ = 0;
   rhs_.outputCol_ = 1;
 }
 
+namespace {
+auto makeInternalVariable(std::string_view string) {
+  return Variable{absl::StrCat("?internal_property_path_variable_", string)};
+}
+}  // namespace
+
+// _____________________________________________________________________________
+std::shared_ptr<QueryExecutionTree> TransitivePathBase::joinWithIndexScan(
+    QueryExecutionContext* qec, Graphs activeGraphs,
+    const TripleComponent& tripleComponent) {
+  auto x = makeInternalVariable("x");
+  auto valuesClause = ad_utility::makeExecutionTree<Values>(
+      qec, parsedQuery::SparqlValues{{x}, {{tripleComponent}}});
+  auto emptyPathSide = makeEmptyPathSide(qec, std::move(activeGraphs));
+  return ad_utility::makeExecutionTree<Join>(qec, std::move(emptyPathSide),
+                                             std::move(valuesClause), 0, 0);
```

Comment:
If you implement it as 
```
{ VALUES + join} UNION {VALUES + join}
```
(that way we will trigger the optimization already today)
then we can merge it without additional penalties.